jolt-core/src/jolt/vm/bytecode.rs

```diff
@@ -343,142 +279,554 @@ impl<F: JoltField, ProofTranscript: Transcript> BytecodeShoutProof<F, ProofTrans
+        let r_cycle_rev: Vec<_> = r_cycle.iter().copied().rev().collect();
 
         let r_concat = [r_address_rev.as_slice(), r_cycle.as_slice()].concat();
         let ra_commitment = &commitments.commitments[CommittedPolynomials::BytecodeRa.to_index()];
-        opening_accumulator.append(&[ra_commitment], r_concat, &[self.ra_claim], transcript);
-
-        let (sumcheck_claim, r_booleanity) =
-            self.booleanity_sumcheck
-                .verify(F::zero(), K.log_2() + T.log_2(), 3, transcript)?;
-        let (r_address_prime, r_cycle_prime) = r_booleanity.split_at(K.log_2());
-        let eq_eval_address = EqPolynomial::mle(&r_address_rev, r_address_prime);
-        let eq_eval_cycle = EqPolynomial::mle(&r_cycle_rev, r_cycle_prime);
-
-        assert_eq!(
-            eq_eval_address * eq_eval_cycle * (self.ra_claim_prime.square() - self.ra_claim_prime),
-            sumcheck_claim,
-            "Booleanity sumcheck failed"
-        );
-
-        let r_address_prime = r_address_prime.iter().copied().rev().collect::<Vec<_>>();
-        let r_cycle_prime = r_cycle_prime.iter().rev().copied().collect::<Vec<_>>();
-        let r_concat = [r_address_prime.as_slice(), r_cycle_prime.as_slice()].concat();
         opening_accumulator.append(
             &[ra_commitment],
             r_concat,
-            &[self.ra_claim_prime],
+            &[self.core_piop_hamming.ra_claim],
             transcript,
         );
 
+        // Verify booleanity sumcheck
+        let (r_booleanity, ra_claim_prime) =
+            self.booleanity
+                .verify(&r_address_rev, &r_cycle_rev, K, T, transcript)?;
+
+        let (r_address_prime, r_cycle_prime) = r_booleanity.split_at(K.log_2());
+        let r_address_prime = r_address_prime.iter().copied().rev().collect::<Vec<_>>();
+        let r_cycle_prime = r_cycle_prime.iter().rev().copied().collect::<Vec<_>>();
+        let r_concat = [r_address_prime.as_slice(), r_cycle_prime.as_slice()].concat();
+
+        opening_accumulator.append(&[ra_commitment], r_concat, &[ra_claim_prime], transcript);
+
         let challenge: F = transcript.challenge_scalar();
         let _ = self.raf_sumcheck.verify(K, challenge, transcript)?;
 
         Ok(())
     }
 }
 
-/// Implements the sumcheck prover for the Booleanity check in step 3 of
-/// Figure 6 in the Twist+Shout paper. The efficient implementation of this
-/// sumcheck is described in Section 6.3.
-#[tracing::instrument(skip_all, name = "Shout booleanity sumcheck")]
-pub fn prove_booleanity<F: JoltField, ProofTranscript: Transcript>(
-    preprocessing: &BytecodePreprocessing,
-    trace: &[RV32IMCycle],
-    r: &[F],
-    D: Vec<F>,
+struct CorePIOPHammingProverState<F: JoltField> {
+    ra_poly: MultilinearPolynomial<F>,
+    val_poly: MultilinearPolynomial<F>,
+    z: F,
+    K: usize,
+}
+
+struct CorePIOPHammingVerifierState<F: JoltField> {
+    z: F,
+    K: usize,
```

Comment:
since `z` and `K` are shared by prover state and verifier state, let's just move them to the `CorePIOPHammingSumcheck` struct and eliminate `CorePIOPHammingVerifierState` entirely